wp/lib/bap_wp/src/precondition.ml

```diff
@@ -314,6 +315,50 @@ let increment_stack_ptr (post : Constr.t) (env : Env.t) (offset : Exp.t)
   let z3_off, _, _, env = exp_to_z3 offset env in
   Constr.substitute_one post sp z3_off, env
 
+let get_vars (t : Sub.t) : Var.Set.t =
+  let visitor =
+    (object inherit [Var.Set.t] Term.visitor
+      method! visit_def def vars =
+        let vars = Var.Set.add vars (Def.lhs def) in
+        let vars = Var.Set.union vars (Def.free_vars def) in
+        vars
+      method! visit_jmp jmp vars =
+        Var.Set.union vars (Jmp.free_vars jmp)
+    end)
+  in
+  visitor#visit_sub t Var.Set.empty
+
+let mk_fun_pred (env : Env.t) (sub : Sub.t) (out_vars : Constr.z3_expr list)
+  : Constr.t * Env.t =
+  let ctx = Env.get_context env in
+  let fun_name = (Sub.name sub) ^ "_pred" in
+  (* FIXME:
+     1. The order of the arguments is probably not guaranteed to be the same
+        across the two different binaries.
+     2. We are currently filtering out mem from the input variables. We might want
+        to add it back, but it causes a great slowdown when Z3 checks the precondition. *)
```

Comment:
`mem` filtering seems very reasonable to me. At best we might be able to point to *specific* regions of `mem` as being arguments, but not the whole thing.

Reply:
How do we determine which regions of `mem` is used as as an argument? I can imagine we can determine what arguments are used from the stack based on what is being pushed at call time, and then gets popped base off of instructions like `#33 := mem[RSP, el]:u64`, but I'm not sure how we determine arguments from the heap.

Reply:
Best left until future work.